perception/sandbox/polygrasp/src/polygrasp/robot_interface.py

```diff
@@ -0,0 +1,245 @@
+                no_position=False,
+                initial_position=curr_joint_pos,
+            )
+            return joint_pos_ikpy[1:-1]
+        except ValueError as e:
+            print(f"Can't find IK solution! {e}")
+            return None
+
+    def gripper_open(self):
+        """Open the gripper. (Assumes Robotiq griper)"""
+        self.gripper.goto(1, 1, 1)
+
+    def gripper_close(self):
+        """Closes the gripper. (Assumes Robotiq gripper)"""
+        self.gripper.goto(0, 1, 1)
+
+    def move_until_success(
+        self, position, orientation, time_to_go, max_attempts=5, success_dist=0.05
+    ):
+        states = []
+        for _ in range(max_attempts):
+            joint_pos = self.ik(position, orientation)
+            states += self.move_to_joint_positions(joint_pos, time_to_go=time_to_go)
+            curr_ee_pos, curr_ee_ori = self.get_ee_pose()
+
+            xyz_diff = torch.linalg.norm(curr_ee_pos - position)
+            ori_diff = (
+                R.from_quat(curr_ee_ori).inv() * R.from_quat(orientation)
+            ).magnitude()
+            log.info(f"Dist to goal: xyz diff {xyz_diff}, ori diff {ori_diff}")
+
+            if (
+                states
+                and states[-1].prev_command_successful
+                and xyz_diff < success_dist
+                and ori_diff < 0.2
+            ):
+                break
+        return states
+
+    def check_feasibility(self, point: np.ndarray):
+        return self.ik(point) is not None
+
+    def select_grasp(
+        self, grasps: graspnetAPI.GraspGroup, num_grasp_choices=5
+    ) -> graspnetAPI.Grasp:
+        with torch.no_grad():
+            feasible_i = []
+            for i, grasp in enumerate(grasps):
+                print(f"checking feasibility {i}/{len(grasps)}")
+
+                if grasp.width > self.gripper_max_width:
+                    continue
+
+                grasp_point, grasp_approach_delta, des_ori_quat = compute_des_pose(
+                    grasp
+                )
+                point_a = grasp_point + self.k_approach * grasp_approach_delta
+                point_b = grasp_point + self.k_grasp * grasp_approach_delta
+
+                if self.check_feasibility(point_a) and self.check_feasibility(point_b):
+                    feasible_i.append(i)
+
+                if len(feasible_i) == num_grasp_choices:
+                    if i >= num_grasp_choices:
+                        print(
+                            f"Kinematically filtered {i + 1 - num_grasp_choices} grasps"
+                            " to get 5 feasible positions"
+                        )
+                    break
+
+            # Choose the grasp closest to the neutral position
+            filtered_grasps = grasps[feasible_i]
+            grasp, i = min_dist_grasp_no_z(self.default_ee_quat, filtered_grasps)
+            log.info(f"Closest grasp to ee ori, within top 5: {i + 1}")
+            return grasp, filtered_grasps
```

Comment:
Nit: wouldn't it be cleaner & more readable to return a list of filtered grasp, and an index indicating which grasp is the chosen one?